File: backend/app/schemas/event.py
from datetime import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel


class EventBase(BaseModel):
    event_name: str
    description: Optional[str] = None
    number_of_days: Optional[int] = None


class EventCreate(EventBase):
    pass


class EventUpdate(BaseModel):
    event_name: Optional[str] = None
    description: Optional[str] = None
    number_of_days: Optional[int] = None


class EventVenueRead(BaseModel):
    venue_id: int
    venue_name: str
    city: str
    capacity: int

    model_config = {"from_attributes": True}


class EventTicketConfigRead(BaseModel):
    config_id: int
    schedule_id: int
    ticket_type: str
    price: Decimal
    max_quantity: int
    remaining_quantity: Optional[int] = None

    model_config = {"from_attributes": True}


class EventSeatRead(BaseModel):
    ticket_id: int
    ticket_config_id: int
    ticket_status: str
    row_label: Optional[str] = None
    col_number: Optional[int] = None
    ticket_type: Optional[str] = None
    price: Optional[Decimal] = None

    model_config = {"from_attributes": True}


class EventDayRead(BaseModel):
    event_day_id: int
    schedule_id: int
    date: datetime
    available_tickets: List[EventSeatRead] = []

    model_config = {"from_attributes": True}


class EventScheduleRead(BaseModel):
    schedule_id: int
    event_id: int
    venue_id: int
    registration_start: Optional[datetime] = None
    registration_end: Optional[datetime] = None
    seat_layout: Optional[str] = None
    venue: EventVenueRead
    ticket_configs: List[EventTicketConfigRead] = []
    event_days: List[EventDayRead] = []

    model_config = {"from_attributes": True}


class EventRead(EventBase):
    event_id: int
    banner_url: Optional[str] = None
    status: str = "ACTIVE"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    schedules: List[EventScheduleRead] = []

    model_config = {"from_attributes": True}


class EventListItem(BaseModel):
    event_id: int
    event_name: str
    description: Optional[str] = None
    number_of_days: Optional[int] = None
    banner_url: Optional[str] = None
    status: str = "ACTIVE"

    model_config = {"from_attributes": True}

File: frontend/src/types/event.ts
export type EventStatus = "ACTIVE" | "CANCELLED" | "INACTIVE" | string;

export type EventSummary = {
  event_id: number;
  event_name: string;
  description?: string | null;
  number_of_days?: number | null;
  banner_url?: string | null;
  status: EventStatus;
};

export type EventVenue = {
  venue_id: number;
  venue_name: string;
  city: string;
  capacity: number;
};

export type EventTicketConfig = {
  config_id: number;
  schedule_id: number;
  ticket_type: string;
  price: string | number;
  max_quantity: number;
  remaining_quantity?: number | null;
};

export type EventSeat = {
  ticket_id: number;
  ticket_config_id: number;
  ticket_status: string;
  row_label?: string | null;
  col_number?: number | null;
  ticket_type?: string | null;
  price?: string | number | null;
};

export type EventDay = {
  event_day_id: number;
  schedule_id: number;
  date: string;
  seats: EventSeat[];
  available_tickets: EventSeat[];
};

export type EventSchedule = {
  schedule_id: number;
  event_id: number;
  venue_id: number;
  registration_start?: string | null;
  registration_end?: string | null;
  seat_layout?: string | null;
  venue: EventVenue;
  ticket_configs: EventTicketConfig[];
  event_days: EventDay[];
};

export type EventRead = EventSummary & {
  created_at?: string | null;
  updated_at?: string | null;
  schedules: EventSchedule[];
};
